Spark_Application/bronze_riot/playerlogs/spark_streaming_kafka_to_s3_v2.py

```diff
@@ -0,0 +1,156 @@
+
+os.environ["PYSPARK_PYTHON"] = sys.executable
+os.environ["PYSPARK_DRIVER_PYTHON"] = sys.executable
+
+"""
+Spark 세션 설정
+"""
+
+spark = (
+    SparkSession.builder.appName("League Of Legend Player Streaming")
+    .master("yarn")
+    .config(
+        "spark.jars.packages",
+        "org.apache.spark:spark-sql-kafka-0-10_2.12:3.3.0,org.apache.spark:spark-streaming-kafka-0-10_2.12:3.3.0",
+    )
+    .config("spark.home", "/usr/lib/spark")
+    .enableHiveSupport()
+    .getOrCreate()
+)
+
+"""
+Kafka 설정
+"""
+kafka_bootstrap_server = sys.argv[1]
+kafka_topic = sys.argv[2]
+
+kafkaStream = (
+    spark.readStream.format("kafka")
+    .option("kafka.bootstrap.servers", kafka_bootstrap_server)
+    .option("subscribe", kafka_topic)
+    .load()
+)
+
+"""
+스키마 정의
+원본 데이터 자체가 String 형이기 때문에 모든 필드를 String Type으로 정의
+"""
+schema = StructType(
+    [
+        StructField("createRoomDate", StringType(), False),
+        StructField("method", StringType(), False),
+        StructField("ingametime", StringType(), False),
+        StructField("ip", StringType(), False),
+        StructField("deathCount", StringType(), False),
+        StructField("roomID", StringType(), False),
+        StructField("datetime", StringType(), False),
+        StructField("x", StringType(), False),
+        StructField("y", StringType(), False),
+        StructField("inputkey", StringType(), False),
+        StructField("account", StringType(), False),
+        StructField("champion", StringType(), False),
+        StructField("status", StringType(), False),
+    ]
+)
+
+"""
+Json 파싱
+"""
+
+jsonParsedStream = (
+    kafkaStream.selectExpr("CAST(value As STRING)")
+    .select(from_json(col("value"), schema).alias("data"))
+    .select("data.*")
+)
+
+
+"""
+적합성 및 유효성 검사
+
+해당 컬럼이 null값을 체크하고, 특정 컬럼은 음수를 가지지 않도록 함
+
+통과한 데이터는 validStream, 통과하지 않는 데이터는 invalidStream
+"""
+validStream = jsonParsedStream.filter(
+    (col("createRoomDate").isNotNull())
+    & (col("ingametime").cast(IntegerType()).isNotNull())
+    & (col("deathCount").cast(IntegerType()).isNotNull())
+    & (col("deathCount").cast(IntegerType()) >= 0)
+    & (col("x").cast(IntegerType()).isNotNull())
+    & (col("y").cast(IntegerType()).isNotNull())
+    & (col("status").cast(IntegerType()).isNotNull())
+    & (col("status").cast(IntegerType()) >= 0)
+)
+
+invalidStream = jsonParsedStream.filter(
+    (col("createRoomDate").isNull())
+    | (col("ingametime").cast(IntegerType()).isNull())
+    | (col("deathCount").cast(IntegerType()).isNull())
+    | (col("deathCount").cast(IntegerType()) <= 0)
+    | (col("x").cast(IntegerType()).isNull())
+    | (col("y").cast(IntegerType()).isNull())
+    | (col("status").cast(IntegerType()).isNull())
+    | (col("status").cast(IntegerType()) <= 0)
+)
```

Comment:
위 내용대로라면 deathCount가 0인 경우에도 Invalid가 맞는지 확인이 필요해보입니다.
그리고 각 검사 조건은 앞으로도 추가, 변경될 수 있을 것으로 보입니다.
가독성과 유지보수성을 높이기 위해 조건을 분리해서 선언하는게 어떨까요?
```suggestion
createroomdate_cond = col("createRoomDate").isNotNull()
ingametime_cond = col("ingametime").cast(IntegerType()).isNotNull()
deathcount_cond = col("deathCount").cast(IntegerType()).isNotNull() & (col("deathCount").cast(IntegerType()) >= 0)
x_cond = col("x").cast(IntegerType()).isNotNull()
y_cond = col("y").cast(IntegerType()).isNotNull()
status_cond = col("status").cast(IntegerType()).isNotNull() & (col("status").cast(IntegerType()) >= 0)

valid_conditions = (
    createroomdate_cond
    & ingametime_cond
    & deathcount_cond
    & x_cond
    & y_cond
    & status_cond
)
validStream = jsonParsedStream.filter(valid_conditions)

invalid_conditions = (
    ~createroomdate_cond
    | ~ingametime_cond
    | ~deathcount_cond
    | ~x_cond
    | ~y_cond
    | ~status_cond
)
invalidStream = jsonParsedStream.filter(invalid_conditions)
```

Reply:
변경했습니다. 정상적으로 음수 값이 아니면 validStream을 통해 최종적으로 bronze_riot/playerlogs 경로에 저장되는 걸 확인했습니다.